test/system/src/test/java/io/pravega/test/system/RetentionTest.java

```diff
@@ -0,0 +1,176 @@
+        Service conService = Utils.createPravegaControllerService(zkUri);
+        if (!conService.isRunning()) {
+            conService.start(true);
+        }
+
+        List<URI> conUris = conService.getServiceDetails();
+        log.debug("Pravega controller service details: {}", conUris);
+
+        //4.start segmentstore
+        Service segService = Utils.createPravegaSegmentStoreService(zkUri, conUris.get(0));
+        if (!segService.isRunning()) {
+            segService.start(true);
+        }
+
+        List<URI> segUris = segService.getServiceDetails();
+        log.debug("Pravega segmentstore service details: {}", segUris);
+    }
+
+    @Before
+    public void setup() {
+        Service conService = Utils.createPravegaControllerService(null);
+        List<URI> ctlURIs = conService.getServiceDetails();
+        controllerURI = ctlURIs.get(0);
+        streamManager = StreamManager.create(controllerURI);
+        assertTrue("Creating Scope", streamManager.createScope(SCOPE));
+        assertTrue("Creating stream", streamManager.createStream(SCOPE, STREAM, config));
+    }
+
+    @Test
+    public void retentionTest() throws Exception {
+
+        ConnectionFactory connectionFactory = new ConnectionFactoryImpl(false);
+        ControllerImpl controller = new ControllerImpl(controllerURI,
+                ControllerImplConfig.builder().build(), connectionFactory.getInternalExecutor());
+
+        ClientFactory clientFactory = new ClientFactoryImpl(SCOPE, controller);
+        log.info("Invoking Writer test with Controller URI: {}", controllerURI);
+
+        //create a writer
+        EventStreamWriter<Serializable> writer = clientFactory.createEventWriter(STREAM,
+                new JavaSerializer<>(),
+                EventWriterConfig.builder().build());
+
+        //write an event
+        String writeEvent = "event";
+        writer.writeEvent("", writeEvent);
+        writer.flush();
+        log.debug("Writing event: {} ", writeEvent);
+
+        //sleep for 4 mins
+        Exceptions.handleInterrupted(() -> Thread.sleep(4 * 60 * 1000));
+
+        //create a reader
+        ReaderGroupManager groupManager = ReaderGroupManager.withScope(SCOPE, controllerURI);
+        groupManager.createReaderGroup(READER_GROUP, ReaderGroupConfig.builder().startingTime(0).build(),
+                Collections.singleton(STREAM));
+        EventStreamReader<String> reader = clientFactory.createReader(UUID.randomUUID().toString(),
+                READER_GROUP,
+                new JavaSerializer<>(),
+                ReaderConfig.builder().build());
+
+        //read the same event
+        try {
+            String readEvent = reader.readNextEvent(6000).getEvent();
+            log.debug("Reading event: {} ", readEvent);
+            assertEquals(null, readEvent);
+        }  catch (ReinitializationRequiredException e) {
+            log.error("Unexpected request to reinitialize {}", e);
+            System.exit(-1);
```

Comment:
We could use an assert.fail() here.

Reply:
done. Failing the test if we get such exception.